File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    bool solve(TreeNode* root, long long low, long long high) {

        // Empty tree is always BST
        if (root == NULL)
            return true;

        // Agar node allowed range ke bahar hai
        if (root->val <= low || root->val >= high)
            return false;

        // Left subtree ka max = root->val
        // Right subtree ka min = root->val
        return solve(root->left, low, root->val) &&
               solve(root->right, root->val, high);
    }

    bool isValidBST(TreeNode* root) {

        return solve(root, LLONG_MIN, LLONG_MAX);
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def solve(temp, low, high):
            if temp == None:
                return True
            if not (low < temp.val < high):
                return False

            return solve(temp.left, low, temp.val) and solve(temp.right, temp.val, high)

        return solve(root, float("-inf"), float("inf"))
